contracts/BondCurve.aes

```diff
@@ -0,0 +1,53 @@
+@compiler >=4.2
+
+include "Frac.aes"
+
+contract BondCurve =
+
+ // token : TokenContract
+  record state = { init_buy_price : Frac.frac
+                 , init_sell_price : Frac.frac
+                 , alpha: Frac.frac }
+
+  stateful entrypoint init() =
+    // { total_supply = 0, init_buy_price = 1, init_sell_price = 0, alpha = 1 }
+    { init_buy_price = Frac.from_int(1), init_sell_price = Frac.from_int(0), alpha = Frac.from_int(1) }
+    
+  // Returns the current buy price
+  entrypoint buy_price(total_supply: Frac.frac) : Frac.frac =
+    buy_curve(total_supply)
+  
+  // Returns the current sell price
+  entrypoint sell_price(total_supply: Frac.frac) : Frac.frac =
+    sell_curve(total_supply)
+  
+  // Returns the amount of fungible tokens to be received
+  // when buying from the curve for the passed value of AE tokens.
+  // entrypoint calculate_buy_return(amount: Frac.frac) : Frac.frac = 0
+    // delta = b^2 + 2a*((a * c^2/2 + b * c) + AE)
+    // n = (-b + sqrt(delta)) / a
+    // Requires SQRT and approximation which is impossible rn
+
+  entrypoint buy_tokens_price(total_supply: Frac.frac, buy_tokens: Frac.frac) : Frac.frac =
+    // (buy_price() + buy_curve(total_supply + buy_tokens)) * buy_tokens / Frac.from_int(2)
+    Frac.div(Frac.mul((Frac.add(buy_price(total_supply), buy_curve(Frac.add(total_supply, buy_tokens)))), buy_tokens), Frac.from_int(2))
```

Comment:
Here the reversed approach of buying tokens is used (this will break the current interface).

It is easier to implement in this way as otherwise, we will need to still use approximations and `sqrt` which is not doable currently in Sophia.

So instead of using the `calculate_buy_return` and passing the `Call.value` to receive the amount of fungible tokens to be received for the passed AE we will use `buy_tokens_price` (this is WIP name).

`buy_tokens_price` will get the desired amount of tokens to be bought by the user and return the amount of AE to be paid.